flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/functions/source/SocketTextStreamFunction.java

```diff
@@ -117,12 +122,13 @@ private void streamFromSocket(SourceContext<String> ctx, Socket socket) throws E
 					continue;
 				}
 
-				if (data == delimiter) {
-					ctx.collect(buffer.toString());
-					buffer = new StringBuilder();
-				} else if (data != '\r') { // ignore carriage return
-					buffer.append((char) data);
+				buffer.append(charBuffer, 0, readCount);
+				String[] splits = buffer.toString().split(delimiter);
```

Comment:
`String.split()` and `String.replace()` create new String objects which must be garbage collected. This adds quite some overhead, because these functions are called very often. The previous implementation was operating on a byte-level and avoiding the creation of new objects. It would be good if we could preserve this behavior.